Windows error: The command is too long

**Describe the bug**
When running a Kotlin script which references many JAR files, the compile steps fails with error "The command is too long".  

**To Reproduce**

```
[jbang] [1:211] Building jar for App.kt...
[jbang] [1:213] Compile: C:\Users\wfouche\.jbang\cache\kotlincs\2.0.21\kotlinc\bin\kotlinc.bat 
-progressive -classpath C:\Users\wfouche\.m2\repository\io\github\wfouche\tulip\tulip-runtime\2.1.8-dev\tulip-runtime-2.1.8-dev.jar;C:\Users\wfouche\.m2\repository\org\hdrhistogram\HdrHistogram\2.2.2\HdrHistogram-2.2.2.jar;C:\Users\wfouche\.m2\repository\org\python\jython-slim\2.7.4\jython-slim-2.7.4.jar;C:\Users\wfouche\.m2\repository\org\antlr\antlr\3.5.3\antlr-3.5.3.jar;C:\Users\wfouche\.m2\repository\org\antlr\ST4\4.3.1\ST4-4.3.1.jar;C:\Users\wfouche\.m2\repository\javax\servlet\javax.servlet-api\3.1.0\javax.servlet-api-3.1.0.jar;C:\Users\wfouche\.m2\repository\org\antlr\antlr-runtime\3.5.3\antlr-runtime-3.5.3.jar;C:\Users\wfouche\.m2\repository\org\apache\commons\commons-compress\1.26.2\commons-compress-1.26.2.jar;C:\Users\wfouche\.m2\repository\commons-codec\commons-codec\1.17.0\commons-codec-1.17.0.jar;C:\Users\wfouche\.m2\repository\org\apache\commons\commons-lang3\3.14.0\commons-lang3-3.14.0.jar;C:\Users\wfouche\.m2\repository\commons-io\commons-io\2.16.1\commons-io-2.16.1.jar;C:\Users\wfouche\.m2\repository\org\bouncycastle\bcpkix-jdk18on\1.78.1\bcpkix-jdk18on-1.78.1.jar;C:\Users\wfouche\.m2\repository\org\bouncycastle\bcutil-jdk18on\1.78.1\bcutil-jdk18on-1.78.1.jar;C:\Users\wfouche\.m2\repository\org\bouncycastle\bcprov-jdk18on\1.78.1\bcprov-jdk18on-1.78.1.jar;C:\Users\wfouche\.m2\repository\org\ow2\asm\asm\9.7\asm-9.7.jar;C:\Users\wfouche\.m2\repository\org\ow2\asm\asm-commons\9.7\asm-commons-9.7.jar;C:\Users\wfouche\.m2\repository\org\ow2\asm\asm-tree\9.7\asm-tree-9.7.jar;C:\Users\wfouche\.m2\repository\org\ow2\asm\asm-util\9.7\asm-util-9.7.jar;C:\Users\wfouche\.m2\repository\org\ow2\asm\asm-analysis\9.7\asm-analysis-9.7.jar;C:\Users\wfouche\.m2\repository\com\google\guava\guava\33.2.1-jre\guava-33.2.1-jre.jar;C:\Users\wfouche\.m2\repository\com\google\guava\failureaccess\1.0.2\failureaccess-1.0.2.jar;C:\Users\wfouche\.m2\repository\com\google\guava\listenablefuture\9999.0-empty-to-avoid-conflict-with-guava\listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar;C:\Users\wfouche\.m2\repository\com\google\code\findbugs\jsr305\3.0.2\jsr305-3.0.2.jar;C:\Users\wfouche\.m2\repository\org\checkerframework\checker-qual\3.42.0\checker-qual-3.42.0.jar;C:\Users\wfouche\.m2\repository\com\google\j2objc\j2objc-annotations\3.0.0\j2objc-annotations-3.0.0.jar;C:\Users\wfouche\.m2\repository\com\ibm\icu\icu4j\75.1\icu4j-75.1.jar;C:\Users\wfouche\.m2\repository\com\carrotsearch\java-sizeof\0.0.5\java-sizeof-0.0.5.jar;C:\Users\wfouche\.m2\repository\com\github\jnr\jffi\1.3.13\jffi-1.3.13.jar;C:\Users\wfouche\.m2\repository\com\github\jnr\jnr-constants\0.10.4\jnr-constants-0.10.4.jar;C:\Users\wfouche\.m2\repository\com\github\jnr\jnr-ffi\2.2.16\jnr-ffi-2.2.16.jar;C:\Users\wfouche\.m2\repository\com\github\jnr\jffi\1.3.13\jffi-1.3.13-native.jar;C:\Users\wfouche\.m2\repository\com\github\jnr\jnr-a64asm\1.0.0\jnr-a64asm-1.0.0.jar;C:\Users\wfouche\.m2\repository\com\github\jnr\jnr-x86asm\1.0.2\jnr-x86asm-1.0.2.jar;C:\Users\wfouche\.m2\repository\com\github\jnr\jnr-netdb\1.2.0\jnr-netdb-1.2.0.jar;C:\Users\wfouche\.m2\repository\com\github\jnr\jnr-posix\3.1.19\jnr-posix-3.1.19.jar;C:\Users\wfouche\.m2\repository\jline\jline\2.14.6\jline-2.14.6.jar;C:\Users\wfouche\.m2\repository\io\netty\netty-buffer\4.1.73.Final\netty-buffer-4.1.73.Final.jar;C:\Users\wfouche\.m2\repository\io\netty\netty-codec\4.1.73.Final\netty-codec-4.1.73.Final.jar;C:\Users\wfouche\.m2\repository\io\netty\netty-common\4.1.73.Final\netty-common-4.1.73.Final.jar;C:\Users\wfouche\.m2\repository\io\netty\netty-handler\4.1.73.Final\netty-handler-4.1.73.Final.jar;C:\Users\wfouche\.m2\repository\io\netty\netty-tcnative-classes\2.0.46.Final\netty-tcnative-classes-2.0.46.Final.jar;C:\Users\wfouche\.m2\repository\io\netty\netty-resolver\4.1.73.Final\netty-resolver-4.1.73.Final.jar;C:\Users\wfouche\.m2\repository\io\netty\netty-transport\4.1.73.Final\netty-transport-4.1.73.Final.jar;C:\Users\wfouche\.m2\repository\org\apache\ant\ant\1.10.14\ant-1.10.14.jar;C:\Users\wfouche\.m2\repository\org\apache\ant\ant-launcher\1.10.14\ant-launcher-1.10.14.jar;C:\Users\wfouche\.m2\repository\org\jetbrains\kotlinx\kotlinx-serialization-json-jvm\1.8.1\kotlinx-serialization-json-jvm-1.8.1.jar;C:\Users\wfouche\.m2\repository\org\jetbrains\kotlinx\kotlinx-serialization-core-jvm\1.8.1\kotlinx-serialization-core-jvm-1.8.1.jar;C:\Users\wfouche\.m2\repository\io\leego\banana\2.1.0\banana-2.1.0.jar;C:\Users\wfouche\.m2\repository\com\google\code\gson\gson\2.13.1\gson-2.13.1.jar;C:\Users\wfouche\.m2\repository\com\google\errorprone\error_prone_annotations\2.38.0\error_prone_annotations-2.38.0.jar;C:\Users\wfouche\.m2\repository\org\apache\httpcomponents\client5\httpclient5\5.5\httpclient5-5.5.jar;C:\Users\wfouche\.m2\repository\org\apache\httpcomponents\core5\httpcore5\5.3.4\httpcore5-5.3.4.jar;C:\Users\wfouche\.m2\repository\org\apache\httpcomponents\core5\httpcore5-h2\5.3.4\httpcore5-h2-5.3.4.jar;C:\Users\wfouche\.m2\repository\org\springframework\boot\spring-boot-starter-web\3.4.4\spring-boot-starter-web-3.4.4.jar;C:\Users\wfouche\.m2\repository\org\springframework\boot\spring-boot-starter\3.4.4\spring-boot-starter-3.4.4.jar;C:\Users\wfouche\.m2\repository\org\springframework\boot\spring-boot\3.4.4\spring-boot-3.4.4.jar;C:\Users\wfouche\.m2\repository\org\springframework\boot\spring-boot-autoconfigure\3.4.4\spring-boot-autoconfigure-3.4.4.jar;C:\Users\wfouche\.m2\repository\org\springframework\boot\spring-boot-starter-logging\3.4.4\spring-boot-starter-logging-3.4.4.jar;C:\Users\wfouche\.m2\repository\org\apache\logging\log4j\log4j-to-slf4j\2.24.3\log4j-to-slf4j-2.24.3.jar;C:\Users\wfouche\.m2\repository\org\apache\logging\log4j\log4j-api\2.24.3\log4j-api-2.24.3.jar;C:\Users\wfouche\.m2\repository\org\slf4j\jul-to-slf4j\2.0.17\jul-to-slf4j-2.0.17.jar;C:\Users\wfouche\.m2\repository\jakarta\annotation\jakarta.annotation-api\2.1.1\jakarta.annotation-api-2.1.1.jar;C:\Users\wfouche\.m2\repository\org\springframework\spring-core\6.2.5\spring-core-6.2.5.jar;C:\Users\wfouche\.m2\repository\org\springframework\spring-jcl\6.2.5\spring-jcl-6.2.5.jar;C:\Users\wfouche\.m2\repository\org\yaml\snakeyaml\2.3\snakeyaml-2.3.jar;C:\Users\wfouche\.m2\repository\org\springframework\boot\spring-boot-starter-json\3.4.4\spring-boot-starter-json-3.4.4.jar;C:\Users\wfouche\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.18.3\jackson-databind-2.18.3.jar;C:\Users\wfouche\.m2\repository\com\fasterxml\jackson\core\jackson-annotations\2.18.3\jackson-annotations-2.18.3.jar;C:\Users\wfouche\.m2\repository\com\fasterxml\jackson\core\jackson-core\2.18.3\jackson-core-2.18.3.jar;C:\Users\wfouche\.m2\repository\com\fasterxml\jackson\datatype\jackson-datatype-jdk8\2.18.3\jackson-datatype-jdk8-2.18.3.jar;C:\Users\wfouche\.m2\repository\com\fasterxml\jackson\datatype\jackson-datatype-jsr310\2.18.3\jackson-datatype-jsr310-2.18.3.jar;C:\Users\wfouche\.m2\repository\com\fasterxml\jackson\module\jackson-module-parameter-names\2.18.3\jackson-module-parameter-names-2.18.3.jar;C:\Users\wfouche\.m2\repository\org\springframework\boot\spring-boot-starter-tomcat\3.4.4\spring-boot-starter-tomcat-3.4.4.jar;C:\Users\wfouche\.m2\repository\org\apache\tomcat\embed\tomcat-embed-core\10.1.39\tomcat-embed-core-10.1.39.jar;C:\Users\wfouche\.m2\repository\org\apache\tomcat\embed\tomcat-embed-el\10.1.39\tomcat-embed-el-10.1.39.jar;C:\Users\wfouche\.m2\repository\org\apache\tomcat\embed\tomcat-embed-websocket\10.1.39\tomcat-embed-websocket-10.1.39.jar;C:\Users\wfouche\.m2\repository\org\springframework\spring-web\6.2.5\spring-web-6.2.5.jar;C:\Users\wfouche\.m2\repository\org\springframework\spring-beans\6.2.5\spring-beans-6.2.5.jar;C:\Users\wfouche\.m2\repository\io\micrometer\micrometer-observation\1.14.5\micrometer-observation-1.14.5.jar;C:\Users\wfouche\.m2\repository\io\micrometer\micrometer-commons\1.14.5\micrometer-commons-1.14.5.jar;C:\Users\wfouche\.m2\repository\org\springframework\spring-webmvc\6.2.5\spring-webmvc-6.2.5.jar;C:\Users\wfouche\.m2\repository\org\springframework\spring-aop\6.2.5\spring-aop-6.2.5.jar;C:\Users\wfouche\.m2\repository\org\springframework\spring-context\6.2.5\spring-context-6.2.5.jar;C:\Users\wfouche\.m2\repository\org\springframework\spring-expression\6.2.5\spring-expression-6.2.5.jar;C:\Users\wfouche\.m2\repository\org\slf4j\slf4j-api\2.0.17\slf4j-api-2.0.17.jar;C:\Users\wfouche\.m2\repository\ch\qos\logback\logback-core\1.5.18\logback-core-1.5.18.jar;C:\Users\wfouche\.m2\repository\ch\qos\logback\logback-classic\1.5.18\logback-classic-1.5.18.jar;C:\Users\wfouche\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib\2.0.21\kotlin-stdlib-2.0.21.jar;C:\Users\wfouche\.m2\repository\org\jetbrains\annotations\13.0\annotations-13.0.jar -d C:\Users\wfouche\.jbang\cache\jars\App.kt.a277408206e614ae2c4ca01f8f5ad3599827a155411d19254e7ea0bcb4a6c436\classes C:\Users\wfouche\IdeaProjects\jbang-catalog\temp\io\tulip\App.kt C:\Users\wfouche\IdeaProjects\jbang-catalog\temp\io\tulip\HttpUser.kt
The command line is too long.
[jbang] [2025-05-26 18:35:27] ProcessBuilder.start(): pid: 16452, dir: null, cmd: "C:\Users\wfouche\.jbang\cache\kotlincs\2.0.21\kotlinc\bin\kotlinc.bat"
[jbang] [1:237] Deleting folder C:\Users\wfouche\.jbang\cache\jars\App.kt.a277408206e614ae2c4ca01f8f5ad3599827a155411d19254e7ea0bcb4a6c436\classes
[jbang] [1:243] Deleting folder C:\Users\wfouche\.jbang\cache\jars\App.kt.a277408206e614ae2c4ca01f8f5ad3599827a155411d19254e7ea0bcb4a6c436\generated
[jbang] [1:244] [ERROR] Error during compile
dev.jbang.cli.ExitException: Error during compile
        at dev.jbang.source.buildsteps.CompileBuildStep.runCompiler(CompileBuildStep.java:151)
        at dev.jbang.source.buildsteps.CompileBuildStep.runCompiler(CompileBuildStep.java:135)
        at dev.jbang.source.buildsteps.CompileBuildStep.compile(CompileBuildStep.java:119)
        at dev.jbang.source.buildsteps.CompileBuildStep.build(CompileBuildStep.java:56)
        at dev.jbang.source.buildsteps.CompileBuildStep.build(CompileBuildStep.java:41)
        at dev.jbang.source.AppBuilder.build(AppBuilder.java:97)
        at dev.jbang.source.AppBuilder.build(AppBuilder.java:22)
        at dev.jbang.cli.Run.doCall(Run.java:88)
        at dev.jbang.cli.BaseCommand.call(BaseCommand.java:145)
        at dev.jbang.cli.BaseCommand.call(BaseCommand.java:21)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
        at dev.jbang.cli.JBang$3.handle(JBang.java:150)
        at dev.jbang.cli.JBang$3.handle(JBang.java:145)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
        at picocli.CommandLine.execute(CommandLine.java:2170)
        at dev.jbang.Main.main(Main.java:23)
[jbang] [1:247] If you believe this a bug in jbang, open an issue at https://github.com/jbangdev/jbang/issues
[jbang] [2025-05-26 18:35:28] Runtime.exit() called with status: 1
```

**Expected behavior**
This is a known issue in the world of Windows. IntelliJ has a workaround that can be enabled which saves all the command arguments and long classpath to a text file which is than passed to a program as "program.exe @arguments.txt"

**JBang version**
0.126.1

**Additional context**
I don't think this is Kotlin specific and the same error can potentially occurr with Java, or Groovy scripts.
java.exe and javac.exe support the shortened comand-line @arguments.txt convention. kotlinc.bat supports it as well.

*java*:

* `@argument files - One or more argument files containing options`

*javac*:

* `@<filename> - Read options and filenames from file`
*kotlinc*:

* `@<argfile> - Read compiler arguments and file paths from the given file.`

*Intellij shortened command-line*

* https://www.baeldung.com/intellij-idea-command-line-is-too-long

**Proposed Solution**

Modify JBang to always use an @arguments file when starting Windows processes.

For example , invoke the Kotlin compiler as:

`....\kotlincs\2.0.21\kotlinc\bin\kotlinc.bat  @arguments.txt`

The following Java test program runs fine on Windows with command  `jbang run test.java`

```java
///usr/bin/env jbang "$0" "$@" ; exit $?

//DEPS org.springframework.boot:spring-boot-starter-web:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-test:3.4.6
//DEPS org.springframework.boot:spring-boot-configuration-processor:3.4.6
//DEPS org.springframework.boot:spring-boot-starter:3.4.6
//DEPS org.springframework.boot:spring-boot-starter:3.4.6                           
//DEPS org.springframework.boot:spring-boot-starter-activemq:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-actuator:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-amqp:3.4.6                      
//DEPS org.springframework.boot:spring-boot-starter-aop:3.4.6                       
//DEPS org.springframework.boot:spring-boot-starter-artemis:3.4.6                   
//DEPS org.springframework.boot:spring-boot-starter-batch:3.4.6                     
//DEPS org.springframework.boot:spring-boot-starter-cache:3.4.6                     
//DEPS org.springframework.boot:spring-boot-starter-data-cassandra:3.4.6            
//DEPS org.springframework.boot:spring-boot-starter-data-cassandra-reactive:3.4.6   
//DEPS org.springframework.boot:spring-boot-starter-data-couchbase:3.4.6            
//DEPS org.springframework.boot:spring-boot-starter-data-couchbase-reactive:3.4.6   
//DEPS org.springframework.boot:spring-boot-starter-data-elasticsearch:3.4.6        
//DEPS org.springframework.boot:spring-boot-starter-data-jdbc:3.4.6                 
//DEPS org.springframework.boot:spring-boot-starter-data-jpa:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-data-ldap:3.4.6                 
//DEPS org.springframework.boot:spring-boot-starter-data-mongodb:3.4.6              
//DEPS org.springframework.boot:spring-boot-starter-data-mongodb-reactive:3.4.6     
//DEPS org.springframework.boot:spring-boot-starter-data-neo4j:3.4.6                
//DEPS org.springframework.boot:spring-boot-starter-data-r2dbc:3.4.6         

import static java.lang.System.*;

public class test {

    public static void main(String... args) {
        out.println("Java: Hello World");
	System.exit(0);
    }
}
```

The equivalent Kotlin program fails however.

```
///usr/bin/env jbang "$0" "$@" ; exit $?

//DEPS org.springframework.boot:spring-boot-starter-web:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-test:3.4.6
//DEPS org.springframework.boot:spring-boot-configuration-processor:3.4.6
//DEPS org.springframework.boot:spring-boot-starter:3.4.6
//DEPS org.springframework.boot:spring-boot-starter:3.4.6                           
//DEPS org.springframework.boot:spring-boot-starter-activemq:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-actuator:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-amqp:3.4.6                      
//DEPS org.springframework.boot:spring-boot-starter-aop:3.4.6                       
//DEPS org.springframework.boot:spring-boot-starter-artemis:3.4.6                   
//DEPS org.springframework.boot:spring-boot-starter-batch:3.4.6                     
//DEPS org.springframework.boot:spring-boot-starter-cache:3.4.6                     
//DEPS org.springframework.boot:spring-boot-starter-data-cassandra:3.4.6            
//DEPS org.springframework.boot:spring-boot-starter-data-cassandra-reactive:3.4.6   
//DEPS org.springframework.boot:spring-boot-starter-data-couchbase:3.4.6            
//DEPS org.springframework.boot:spring-boot-starter-data-couchbase-reactive:3.4.6   
//DEPS org.springframework.boot:spring-boot-starter-data-elasticsearch:3.4.6        
//DEPS org.springframework.boot:spring-boot-starter-data-jdbc:3.4.6                 
//DEPS org.springframework.boot:spring-boot-starter-data-jpa:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-data-ldap:3.4.6                 
//DEPS org.springframework.boot:spring-boot-starter-data-mongodb:3.4.6              
//DEPS org.springframework.boot:spring-boot-starter-data-mongodb-reactive:3.4.6     
//DEPS org.springframework.boot:spring-boot-starter-data-neo4j:3.4.6                
//DEPS org.springframework.boot:spring-boot-starter-data-r2dbc:3.4.6         

fun main(args: Array<String>) {
    println("Kotlin: Hello World");
}
```

```
[jbang] Building jar for test.kt...
The command line is too long.
[jbang] [ERROR] Error during compile
[jbang] Run with --verbose for more details. The --verbose must be placed before the jbang command. I.e. jbang --verbose run [...]
```

The equivalent Groovy script also fails.

```
///usr/bin/env jbang "$0" "$@" ; exit $?

//DEPS org.springframework.boot:spring-boot-starter-web:3.4.6
//DEPS org.springframework.boot:spring-boot-starter-test:3.4.6
//DEPS org.springframework.boot:spring-boot-configuration-processor:3.4.6
//DEPS org.springframework.boot:spring-boot-starter:3.4.6
//DEPS org.springframework.boot:spring-boot-starter:3.4.6                           
//DEPS org.springframework.boot:spring-boot-starter-activemq:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-actuator:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-amqp:3.4.6                      
//DEPS org.springframework.boot:spring-boot-starter-aop:3.4.6                       
//DEPS org.springframework.boot:spring-boot-starter-artemis:3.4.6                   
//DEPS org.springframework.boot:spring-boot-starter-batch:3.4.6                     
//DEPS org.springframework.boot:spring-boot-starter-cache:3.4.6                     
//DEPS org.springframework.boot:spring-boot-starter-data-cassandra:3.4.6            
//DEPS org.springframework.boot:spring-boot-starter-data-cassandra-reactive:3.4.6   
//DEPS org.springframework.boot:spring-boot-starter-data-couchbase:3.4.6            
//DEPS org.springframework.boot:spring-boot-starter-data-couchbase-reactive:3.4.6   
//DEPS org.springframework.boot:spring-boot-starter-data-elasticsearch:3.4.6        
//DEPS org.springframework.boot:spring-boot-starter-data-jdbc:3.4.6                 
//DEPS org.springframework.boot:spring-boot-starter-data-jpa:3.4.6                  
//DEPS org.springframework.boot:spring-boot-starter-data-ldap:3.4.6                 
//DEPS org.springframework.boot:spring-boot-starter-data-mongodb:3.4.6              
//DEPS org.springframework.boot:spring-boot-starter-data-mongodb-reactive:3.4.6     
//DEPS org.springframework.boot:spring-boot-starter-data-neo4j:3.4.6                
//DEPS org.springframework.boot:spring-boot-starter-data-r2dbc:3.4.6 

println("Hello World");
```

```
jbang] Building jar [for test.groovy...
The command line is too long.
[jbang] [ERROR] Error during compile
[jbang] Run with --verbose for more details. The --verbose must be placed before the jbang command. I.e. jbang --verbose run [...]
```

--- a/src/main/java/dev/jbang/cli/App.java
+++ b/src/main/java/dev/jbang/cli/App.java
@@ -178,7 +178,7 @@ class AppInstall extends BaseCommand {
 		cmd.add(scriptRef);
 		cmd.addAll(runArgs);
 		CommandBuffer cb = CommandBuffer.of(cmd);
-		List<String> lines = Arrays.asList("#!/bin/sh", cb.asCommandLine(Util.Shell.bash) + " \"$@\"");
+		List<String> lines = Arrays.asList("#!/bin/sh", cb.shell(Util.Shell.bash).asCommandLine() + " \"$@\"");
 		Files.write(file, lines, StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.CREATE);
 		if (!Util.isWindows()) {
 			Util.setExecutable(file);
@@ -193,7 +193,7 @@ class AppInstall extends BaseCommand {
 		cmd.add(scriptRef);
 		cmd.addAll(runArgs);
 		CommandBuffer cb = CommandBuffer.of(cmd);
-		List<String> lines = Arrays.asList("@echo off", cb.asCommandLine(Util.Shell.cmd) + " %*");
+		List<String> lines = Arrays.asList("@echo off", cb.shell(Util.Shell.cmd).asCommandLine() + " %*");
 		Files.write(file, lines, StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.CREATE);
 	}
 
@@ -205,7 +205,7 @@ class AppInstall extends BaseCommand {
 		cmd.add(scriptRef);
 		cmd.addAll(runArgs);
 		CommandBuffer cb = CommandBuffer.of(cmd);
-		List<String> lines = Collections.singletonList(cb.asCommandLine(Util.Shell.powershell) + " @args");
+		List<String> lines = Collections.singletonList(cb.shell(Util.Shell.powershell).asCommandLine() + " @args");
 		Files.write(file, lines, StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.CREATE);
 	}
 

--- a/src/main/java/dev/jbang/cli/Edit.java
+++ b/src/main/java/dev/jbang/cli/Edit.java
@@ -261,10 +261,10 @@ public class Edit extends BaseCommand {
 
 			String[] cmd;
 			if (Util.getShell() == Shell.bash) {
-				final String editorCommand = CommandBuffer.of(optionList).asCommandLine(Shell.bash);
+				final String editorCommand = CommandBuffer.of(optionList).shell(Shell.bash).asCommandLine();
 				cmd = new String[] { "sh", "-c", editorCommand };
 			} else {
-				final String editorCommand = CommandBuffer.of(optionList).asCommandLine(Shell.cmd);
+				final String editorCommand = CommandBuffer.of(optionList).shell(Shell.cmd).asCommandLine();
 				cmd = new String[] { "cmd", "/c", editorCommand };
 			}
 			verboseMsg("Running `" + String.join(" ", cmd) + "`");

--- a/src/main/java/dev/jbang/source/buildsteps/CompileBuildStep.java
+++ b/src/main/java/dev/jbang/source/buildsteps/CompileBuildStep.java
@@ -133,8 +133,7 @@ public abstract class CompileBuildStep implements Builder<Project> {
 
 	protected void runCompiler(List<String> optionList) throws IOException {
 		runCompiler(CommandBuffer.of(optionList)
-			.applyWindowsMaxLengthLimit(CommandBuffer.MAX_LENGTH_WINPROCBUILDER,
-					Util.getShell())
+			.applyWindowsMaxLengthLimit()
 			.asProcessBuilder()
 			.inheritIO());
 	}

--- a/src/main/java/dev/jbang/source/buildsteps/CompileBuildStep.java
+++ b/src/main/java/dev/jbang/source/buildsteps/CompileBuildStep.java
@@ -6,7 +6,6 @@ import java.io.InputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
 import java.util.Optional;
@@ -23,7 +22,6 @@ import org.jboss.jandex.Type;
 import dev.jbang.cli.BaseCommand;
 import dev.jbang.cli.ExitException;
 import dev.jbang.dependencies.MavenCoordinate;
-import dev.jbang.devkitman.Jdk;
 import dev.jbang.source.BuildContext;
 import dev.jbang.source.Builder;
 import dev.jbang.source.Project;
@@ -57,38 +55,33 @@ public abstract class CompileBuildStep implements Builder<Project> {
 	}
 
 	protected Project compile() throws IOException {
-		Project project = ctx.getProject();
-		Jdk jdk = project.projectJdk();
-		String requestedJavaVersion = project.getJavaVersion();
-		if (requestedJavaVersion == null
-				&& project.getModuleName().isPresent()
-				&& jdk.majorVersion() < 9) {
-			// Make sure we use at least Java 9 when dealing with modules
-			requestedJavaVersion = "9+";
-			jdk = project.projectJdkManager().getOrInstallJdk(requestedJavaVersion);
-		}
+		List<String> compileCmd = getCompileCommand();
 
-		Path compileDir = ctx.getCompileDir();
-		List<String> optionList = new ArrayList<>();
-		optionList.add(getCompilerBinary(requestedJavaVersion));
-		if (project.enablePreview()) {
-			optionList.add("--enable-preview");
-			optionList.add("-source");
-			optionList.add("" + jdk.majorVersion());
-		}
-		optionList.addAll(project.getMainSourceSet().getCompileOptions());
-		String path = ctx.resolveClassPath().getClassPath();
-		if (!Util.isBlankString(path)) {
-			if (project.getModuleName().isPresent()) {
-				optionList.addAll(Arrays.asList("-p", path));
-			} else {
-				optionList.addAll(Arrays.asList("-classpath", path));
-			}
-		}
-		optionList.addAll(Arrays.asList("-d", compileDir.toAbsolutePath().toString()));
+		// add additional files
+		Project project = ctx.getProject();
+		project.getMainSourceSet().copyResourcesTo(ctx.getCompileDir());
+
+		generatePom();
+
+		Util.infoMsg(String.format("Building %s for %s...", project.getMainSource().isAgent() ? "javaagent" : "jar",
+				project.getResourceRef().getFile().getFileName().toString()));
+		Util.verboseMsg("Compile: " + String.join(" ", compileCmd));
+		runCompiler(compileCmd);
+
+		searchForMain(ctx.getCompileDir());
+
+		return project;
+	}
+
+	protected List<String> getCompileCommand() throws IOException {
+		List<String> compileCmd = new ArrayList<>();
+
+		Project project = ctx.getProject();
+		compileCmd.add(getCompilerBinary());
+		compileCmd.addAll(getCompileCommandOptions());
 
 		// add source files to compile
-		optionList.addAll(project.getMainSourceSet()
+		compileCmd.addAll(project.getMainSourceSet()
 			.getSources()
 			.stream()
 			.map(x -> x.getFile().toString())
@@ -103,24 +96,12 @@ public abstract class CompileBuildStep implements Builder<Project> {
 				// generate module-info descriptor and add it to list of files to compile
 				Path infoFile = ModuleUtil.generateModuleInfo(ctx);
 				if (infoFile != null) {
-					optionList.add(infoFile.toString());
+					compileCmd.add(infoFile.toString());
 				}
 			}
 		}
 
-		// add additional files
-		project.getMainSourceSet().copyResourcesTo(compileDir);
-
-		generatePom();
-
-		Util.infoMsg(String.format("Building %s for %s...", project.getMainSource().isAgent() ? "javaagent" : "jar",
-				project.getResourceRef().getFile().getFileName().toString()));
-		Util.verboseMsg("Compile: " + String.join(" ", optionList));
-		runCompiler(optionList);
-
-		searchForMain(compileDir);
-
-		return project;
+		return compileCmd;
 	}
 
 	private boolean hasModuleInfoFile() {
@@ -130,6 +111,10 @@ public abstract class CompileBuildStep implements Builder<Project> {
 			.stream()
 			.anyMatch(s -> s.getFile().getFileName().toString().equals("module-info.java"));
 	}
+
+	protected abstract String getCompilerBinary();
+
+	protected abstract List<String> getCompileCommandOptions() throws IOException;
 
 	protected void runCompiler(List<String> optionList) throws IOException {
 		runCompiler(CommandBuffer.of(optionList)
@@ -285,13 +270,10 @@ public abstract class CompileBuildStep implements Builder<Project> {
 		}
 	}
 
-	protected abstract String getCompilerBinary(String requestedJavaVersion);
-
 	protected abstract String getMainExtension();
 
 	protected Predicate<ClassInfo> getMainFinder() {
 		return pubClass -> (pubClass.method("main", STRINGARRAYTYPE) != null
-				||
-				pubClass.method("main") != null);
+				|| pubClass.method("main") != null);
 	}
 }

--- a/src/main/java/dev/jbang/source/buildsteps/NativeBuildStep.java
+++ b/src/main/java/dev/jbang/source/buildsteps/NativeBuildStep.java
@@ -72,7 +72,7 @@ public class NativeBuildStep implements Builder<Project> {
 		Util.verboseMsg("native-image: " + String.join(" ", optionList));
 
 		ProcessBuilder pb = CommandBuffer.of(optionList)
-			.applyWindowsMaxLengthLimit(32000, Util.getShell())
+			.applyWindowsMaxLengthLimit()
 			.asProcessBuilder()
 			.inheritIO();
 

--- a/src/main/java/dev/jbang/source/generators/BaseCmdGenerator.java
+++ b/src/main/java/dev/jbang/source/generators/BaseCmdGenerator.java
@@ -56,6 +56,6 @@ public abstract class BaseCmdGenerator<T extends CmdGenerator> implements CmdGen
 
 	protected String generateCommandLineString(List<String> fullArgs) throws IOException {
 		CommandBuffer cb = CommandBuffer.of(fullArgs);
-		return cb.asCommandLine(shell);
+		return cb.shell(shell).asCommandLine();
 	}
 }

--- a/src/main/java/dev/jbang/source/generators/JarCmdGenerator.java
+++ b/src/main/java/dev/jbang/source/generators/JarCmdGenerator.java
@@ -199,8 +199,9 @@ public class JarCmdGenerator extends BaseCmdGenerator<JarCmdGenerator> {
 
 	protected String generateCommandLineString(List<String> fullArgs) throws IOException {
 		return CommandBuffer.of(fullArgs)
-			.applyWindowsMaxLengthLimit(CommandBuffer.MAX_LENGTH_WINCLI, shell)
-			.asCommandLine(shell);
+			.shell(shell)
+			.applyWindowsMaxLengthLimit(CommandBuffer.MAX_LENGTH_WINCLI)
+			.asCommandLine();
 	}
 
 	private static void addPropertyFlags(Map<String, String> properties, String def, List<String> result) {

--- a/src/main/java/dev/jbang/source/sources/GroovySource.java
+++ b/src/main/java/dev/jbang/source/sources/GroovySource.java
@@ -3,14 +3,13 @@ package dev.jbang.source.sources;
 import static dev.jbang.net.GroovyManager.resolveInGroovyHome;
 
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.function.Function;
-import java.util.function.Predicate;
 
 import javax.annotation.Nonnull;
-
-import org.jboss.jandex.ClassInfo;
 
 import dev.jbang.net.GroovyManager;
 import dev.jbang.source.*;
@@ -90,9 +89,17 @@ public class GroovySource extends Source {
 			}
 
 			@Override
-			protected String getCompilerBinary(String requestedJavaVersion) {
+			protected String getCompilerBinary() {
 				return resolveInGroovyHome("groovyc",
 						((GroovySource) ctx.getProject().getMainSource()).getGroovyVersion());
+			}
+
+			@Override
+			protected List<String> getCompileCommandOptions() throws IOException {
+				List<String> optionList = new ArrayList<>();
+				optionList.addAll(ctx.getProject().getMainSourceSet().getCompileOptions());
+				optionList.addAll(Arrays.asList("-d", ctx.getCompileDir().toAbsolutePath().toString()));
+				return optionList;
 			}
 
 			@Override
@@ -105,6 +112,10 @@ public class GroovySource extends Source {
 									.home()
 									.toString());
 					processBuilder.environment().remove("GROOVY_HOME");
+					String path = ctx.resolveClassPath().getClassPath();
+					if (!Util.isBlankString(path)) {
+						processBuilder.environment().put("CLASSPATH", path);
+					}
 				}
 				super.runCompiler(processBuilder);
 			}
@@ -112,12 +123,6 @@ public class GroovySource extends Source {
 			@Override
 			protected String getMainExtension() {
 				return Type.groovy.extension;
-			}
-
-			@Override
-			protected Predicate<ClassInfo> getMainFinder() {
-				return pubClass -> pubClass.method("main", CompileBuildStep.STRINGARRAYTYPE) != null
-						|| pubClass.method("main") != null;
 			}
 		}
 	}

--- a/src/main/java/dev/jbang/source/sources/JavaSource.java
+++ b/src/main/java/dev/jbang/source/sources/JavaSource.java
@@ -2,6 +2,9 @@ package dev.jbang.source.sources;
 
 import static dev.jbang.util.JavaUtil.resolveInJavaHome;
 
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.function.Function;
 
@@ -11,6 +14,7 @@ import dev.jbang.devkitman.Jdk;
 import dev.jbang.source.*;
 import dev.jbang.source.AppBuilder;
 import dev.jbang.source.buildsteps.CompileBuildStep;
+import dev.jbang.util.Util;
 
 public class JavaSource extends Source {
 
@@ -68,10 +72,42 @@ public class JavaSource extends Source {
 			}
 
 			@Override
-			protected String getCompilerBinary(String requestedJavaVersion) {
-				Project prj = ctx.getProject();
-				Jdk jdk = prj.projectJdkManager().getOrInstallJdk(requestedJavaVersion);
+			protected String getCompilerBinary() {
+				Project project = ctx.getProject();
+				Jdk jdk = project.projectJdk();
+				String requestedJavaVersion = project.getJavaVersion();
+				if (requestedJavaVersion == null
+						&& project.getModuleName().isPresent()
+						&& jdk.majorVersion() < 9) {
+					// Make sure we use at least Java 9 when dealing with modules
+					requestedJavaVersion = "9+";
+					jdk = project.projectJdkManager().getOrInstallJdk(requestedJavaVersion);
+				}
 				return resolveInJavaHome("javac", jdk);
+			}
+
+			@Override
+			protected List<String> getCompileCommandOptions() throws IOException {
+				List<String> optionList = new ArrayList<>();
+
+				Project project = ctx.getProject();
+				if (project.enablePreview()) {
+					optionList.add("--enable-preview");
+					optionList.add("-source");
+					optionList.add("" + project.projectJdk().majorVersion());
+				}
+				optionList.addAll(project.getMainSourceSet().getCompileOptions());
+				String path = ctx.resolveClassPath().getClassPath();
+				if (!Util.isBlankString(path)) {
+					if (project.getModuleName().isPresent()) {
+						optionList.addAll(Arrays.asList("-p", path));
+					} else {
+						optionList.addAll(Arrays.asList("-classpath", path));
+					}
+				}
+				optionList.addAll(Arrays.asList("-d", ctx.getCompileDir().toAbsolutePath().toString()));
+
+				return optionList;
 			}
 
 			@Override

--- a/src/main/java/dev/jbang/source/sources/KotlinSource.java
+++ b/src/main/java/dev/jbang/source/sources/KotlinSource.java
@@ -2,18 +2,19 @@ package dev.jbang.source.sources;
 
 import static dev.jbang.net.KotlinManager.resolveInKotlinHome;
 
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.function.Function;
-import java.util.function.Predicate;
 
 import javax.annotation.Nonnull;
-
-import org.jboss.jandex.ClassInfo;
 
 import dev.jbang.net.KotlinManager;
 import dev.jbang.source.*;
 import dev.jbang.source.AppBuilder;
 import dev.jbang.source.buildsteps.CompileBuildStep;
+import dev.jbang.util.Util;
 
 public class KotlinSource extends Source {
 
@@ -81,20 +82,26 @@ public class KotlinSource extends Source {
 			}
 
 			@Override
-			protected String getCompilerBinary(String requestedJavaVersion) {
+			protected String getCompilerBinary() {
 				return resolveInKotlinHome("kotlinc",
 						((KotlinSource) ctx.getProject().getMainSource()).getKotlinVersion());
 			}
 
 			@Override
-			protected String getMainExtension() {
-				return Type.kotlin.extension;
+			protected List<String> getCompileCommandOptions() throws IOException {
+				List<String> optionList = new ArrayList<>();
+				optionList.addAll(ctx.getProject().getMainSourceSet().getCompileOptions());
+				String path = ctx.resolveClassPath().getClassPath();
+				if (!Util.isBlankString(path)) {
+					optionList.addAll(Arrays.asList("-classpath", path));
+				}
+				optionList.addAll(Arrays.asList("-d", ctx.getCompileDir().toAbsolutePath().toString()));
+				return optionList;
 			}
 
 			@Override
-			protected Predicate<ClassInfo> getMainFinder() {
-				return pubClass -> pubClass.method("main", CompileBuildStep.STRINGARRAYTYPE) != null
-						|| pubClass.method("main") != null;
+			protected String getMainExtension() {
+				return Type.kotlin.extension;
 			}
 		}
 	}

--- a/src/main/java/dev/jbang/util/CommandBuffer.java
+++ b/src/main/java/dev/jbang/util/CommandBuffer.java
@@ -13,6 +13,7 @@ import java.util.stream.Collectors;
 
 public class CommandBuffer {
 	private List<String> arguments;
+	private Util.Shell shell = Util.getShell();
 
 	// 8192 character command line length limit imposed by CMD.EXE
 	public static final int MAX_LENGTH_WINCLI = 8000;
@@ -52,11 +53,12 @@ public class CommandBuffer {
 		this.arguments = new ArrayList<>(Arrays.asList(arguments));
 	}
 
-	public ProcessBuilder asProcessBuilder() {
-		return asProcessBuilder(Util.getShell());
+	public CommandBuffer shell(Util.Shell shell) {
+		this.shell = shell;
+		return this;
 	}
 
-	public ProcessBuilder asProcessBuilder(Util.Shell shell) {
+	public ProcessBuilder asProcessBuilder() {
 		List<String> args = arguments.stream()
 			.map(a -> escapeProcessBuilderArgument(a, shell))
 			.collect(Collectors.toList());
@@ -64,14 +66,13 @@ public class CommandBuffer {
 	}
 
 	public String asCommandLine() {
-		return asCommandLine(Util.getShell());
-	}
-
-	public String asCommandLine(Util.Shell shell) {
 		return String.join(" ", escapeShellArguments(arguments, shell));
 	}
 
 	public CommandBuffer usingArgsFile() throws IOException {
+		if (arguments.size() < 2 || arguments.get(1).startsWith("@")) {
+			return this;
+		}
 		// @-files avoid problems on Windows with very long command lines
 		final Path argsFile = Files.createTempFile("jbang", ".args");
 		try (PrintWriter pw = new PrintWriter(argsFile.toFile())) {
@@ -83,10 +84,19 @@ public class CommandBuffer {
 		return CommandBuffer.of(arguments.get(0), "@" + argsFile);
 	}
 
-	public CommandBuffer applyWindowsMaxLengthLimit(int maxLength, Util.Shell shell) throws IOException {
-		String args = asCommandLine(shell);
+	public CommandBuffer applyWindowsMaxLengthLimit() throws IOException {
+		int maxLength = MAX_LENGTH_WINPROCBUILDER;
+		String cmd = arguments.get(0).toLowerCase();
+		if (cmd.endsWith(".bat") || cmd.endsWith(".cmd")) {
+			maxLength = MAX_LENGTH_WINCLI;
+		}
+		return applyWindowsMaxLengthLimit(maxLength);
+	}
+
+	public CommandBuffer applyWindowsMaxLengthLimit(int maxLength) throws IOException {
+		String args = asCommandLine();
 		// Check if we can and need to use @-files on Windows
-		if (args.length() > maxLength && Util.getShell() != Util.Shell.bash) {
+		if (args.length() > maxLength && shell != Util.Shell.bash) {
 			return usingArgsFile();
 		} else {
 			return this;

--- a/src/test/java/dev/jbang/util/TestCommandBuffer.java
+++ b/src/test/java/dev/jbang/util/TestCommandBuffer.java
@@ -4,42 +4,102 @@ import static dev.jbang.util.Util.JBANG_RUNTIME_SHELL;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.*;
 
+import java.io.IOException;
+
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledOnOs;
+import org.junit.jupiter.api.condition.OS;
 
 import dev.jbang.BaseTest;
 
 public class TestCommandBuffer extends BaseTest {
 
 	@Test
+	@EnabledOnOs(OS.WINDOWS)
 	void testRunWinBat() {
-		if (Util.getOS() == Util.OS.windows) {
-			String out = Util.runCommand(examplesTestFolder.resolve("echo.bat").toString(), "abc def", "abc;def",
-					"abc=def", "abc,def");
-			assertThat(out, containsString("ARG = abc def"));
-			assertThat(out, containsString("ARG = abc;def"));
-			assertThat(out, containsString("ARG = abc=def"));
-			assertThat(out, containsString("ARG = abc,def"));
-		}
+		String out = Util.runCommand(examplesTestFolder.resolve("echo.bat").toString(), "abc def", "abc;def",
+				"abc=def", "abc,def");
+		assertThat(out, containsString("ARG = abc def"));
+		assertThat(out, containsString("ARG = abc;def"));
+		assertThat(out, containsString("ARG = abc=def"));
+		assertThat(out, containsString("ARG = abc,def"));
 	}
 
 	@Test
+	@EnabledOnOs(OS.WINDOWS)
 	void testRunWinBatFromBash() {
-		if (Util.getOS() == Util.OS.windows) {
-			environmentVariables.set(JBANG_RUNTIME_SHELL, "bash");
-			String out = Util.runCommand(examplesTestFolder.resolve("echo.bat").toString(), "abc def", "abc;def",
-					"abc=def", "abc,def");
-			assertThat(out, containsString("ARG = abc def"));
-			assertThat(out, containsString("ARG = abc;def"));
-			assertThat(out, containsString("ARG = abc=def"));
-			assertThat(out, containsString("ARG = abc,def"));
-		}
+		environmentVariables.set(JBANG_RUNTIME_SHELL, "bash");
+		String out = Util.runCommand(examplesTestFolder.resolve("echo.bat").toString(), "abc def", "abc;def",
+				"abc=def", "abc,def");
+		assertThat(out, containsString("ARG = abc def"));
+		assertThat(out, containsString("ARG = abc;def"));
+		assertThat(out, containsString("ARG = abc=def"));
+		assertThat(out, containsString("ARG = abc,def"));
 	}
 
 	@Test
+	@EnabledOnOs(OS.WINDOWS)
 	void testRunWinPS1() {
-		if (Util.getOS() == Util.OS.windows) {
-			String out = CommandBuffer.of("abc def", "abc;def").asCommandLine(Util.Shell.powershell);
-			assertThat(out, equalTo("'abc def' 'abc;def'"));
+		String out = CommandBuffer.of("abc def", "abc;def").shell(Util.Shell.powershell).asCommandLine();
+		assertThat(out, equalTo("'abc def' 'abc;def'"));
+	}
+
+	@Test
+	void testApplyWindowsMaxLengthLimitExe() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.exe"))
+			.shell(Util.Shell.cmd)
+			.applyWindowsMaxLengthLimit()
+			.asProcessBuilder();
+		assertThat(pb.command().size(), greaterThan(2));
+		assertThat(pb.command().get(1), not(anyOf(startsWith("@"), startsWith("\"@"))));
+	}
+
+	@Test
+	void testApplyWindowsMaxLengthLimitBat() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.bat"))
+			.shell(Util.Shell.cmd)
+			.applyWindowsMaxLengthLimit()
+			.asProcessBuilder();
+		assertThat(pb.command().size(), equalTo(2));
+		assertThat(pb.command().get(1), anyOf(startsWith("@"), startsWith("\"@")));
+	}
+
+	@Test
+	void testApplyWindowsMaxLengthLimitCmd() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of(argsTooLong("foo.cmd"))
+			.shell(Util.Shell.cmd)
+			.applyWindowsMaxLengthLimit()
+			.asProcessBuilder();
+		assertThat(pb.command().size(), equalTo(2));
+		assertThat(pb.command().get(1), anyOf(startsWith("@"), startsWith("\"@")));
+	}
+
+	@Test
+	void testUsingArgsFileWith1Arg() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of("abc").usingArgsFile().asProcessBuilder();
+		assertThat(pb.command().size(), equalTo(1));
+	}
+
+	@Test
+	void testUsingArgsFileWith3Args() throws IOException {
+		ProcessBuilder pb = CommandBuffer.of("abc", "def", "ghi").usingArgsFile().asProcessBuilder();
+		assertThat(pb.command().size(), equalTo(2));
+		assertThat(pb.command().get(1), anyOf(startsWith("@"), startsWith("\"@")));
+	}
+
+	@Test
+	void testUsingArgsFileNoDup() throws IOException {
+		CommandBuffer cmd = CommandBuffer.of("abc", "def", "ghi").usingArgsFile();
+		CommandBuffer cmd2 = cmd.usingArgsFile();
+		assertThat(cmd, is(cmd2));
+	}
+
+	private String[] argsTooLong(String cmd) {
+		String[] args = new String[1000];
+		args[0] = cmd;
+		for (int i = 1; i < args.length; i++) {
+			args[i] = "argument " + i;
 		}
+		return args;
 	}
 }

